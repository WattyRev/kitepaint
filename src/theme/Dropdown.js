import React from "react";
import PropTypes from "prop-types";
import styled from "styled-components";

const StyleWrapper = styled.div`
  position: relative;
`;

/**
 * Applies styles to the dropdown itself
 */
export const StyledDropdown = styled.div`
  position: absolute;
  top: 100%;
  right: 0;
  background: ${props => props.theme.colors.white};
  border-radius: 4px;
  border-top-right-radius: 0;
  border: 1px solid ${props => props.theme.colors.silver};
  text-align: right;
`;

/**
 * Wraps the provided children in a ralative container and allows for opening and closing an
 * absolutely positionined dropdown.
 *
 * Example:
 * ```
 * <Dropdown
 *   dropdownContent={dropdownData => <a onClick={dropdownData.actions.closeDropdown}>Close</a>}
 * >
 *   {dropdownData => <a onClick={dropdownData.actions.openDropdown}>Open</a>}
 * </Dropdown>
 * ```
 */
class Dropdown extends React.Component {
  static propTypes = {
    /**
     * The persistent content that the dropdown is bound to.
     */
    children: PropTypes.func.isRequired,
    /**
     * The contents of the dropdown
     */
    dropdownContent: PropTypes.func.isRequired
  };
  state = {
    /**
     * Indicates if the dropdown is open or not
     */
    isOpen: false
  };

  /**
   * Based on the provided event, close the dropdown if the event happened outside this component.
   * This enables the dropdown to automatically close when clicking outside of it.
   * @param {Object} event A DOM click event
   */
  autoCloseDropdown = event => {
    if (!this.state.isOpen) {
      return;
    }
    if (this.node.contains(event.target)) {
      return;
    }
    this.closeDropdown();
  };

  componentDidMount() {
    document.addEventListener("click", this.autoCloseDropdown);
  }

  componentWillUnmount() {
    document.removeEventListener("click", this.autoCloseDropdown);
  }

  /**
   * Opens the dropdown
   */
  openDropdown = () => {
    this.setState({
      isOpen: true
    });
  };

  /**
   * Closes the dropdown
   */
  closeDropdown = () => {
    this.setState({
      isOpen: false
    });
  };

  render() {
    const renderData = {
      actions: {
        openDropdown: this.openDropdown,
        closeDropdown: this.closeDropdown
      },
      props: {
        isOpen: this.state.isOpen
      }
    };
    return (
      <StyleWrapper ref={node => (this.node = node)}>
        {this.props.children(renderData)}
        {this.state.isOpen && (
          <StyledDropdown>
            {this.props.dropdownContent(renderData)}
          </StyledDropdown>
        )}
      </StyleWrapper>
    );
  }
}

export default Dropdown;
