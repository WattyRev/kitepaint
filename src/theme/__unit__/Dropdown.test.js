import React from "react";
import { shallow, mount } from "enzyme";
import Theme from "../../theme";
import Dropdown, { StyledDropdown, Item, Spacer } from "../Dropdown";

describe("Dropdown", () => {
  it("renders", () => {
    shallow(
      <Dropdown dropdownContent={() => <div>test</div>}>
        {() => <div>test2</div>}
      </Dropdown>
    );
  });
  it("renders when using the provided components", () => {
    shallow(
      <Dropdown dropdownContent={() => <div>test</div>}>
        {data => (
          <React.Fragment>
            <data.components.Item>Item</data.components.Item>
            <data.components.Spacer />
          </React.Fragment>
        )}
      </Dropdown>
    );
  });
  it("renders the dropdown content when open is triggered", () => {
    expect.assertions(2);
    const wrapper = shallow(
      <Dropdown dropdownContent={() => <div className="target">test</div>}>
        {data => (
          <div className="click-open" onClick={data.actions.open}>
            test2
          </div>
        )}
      </Dropdown>
    );
    expect(wrapper.find(".target")).toHaveLength(0);
    wrapper.find(".click-open").simulate("click");
    expect(wrapper.find(".target")).toHaveLength(1);
  });
  it("removes the dropdown content when close is triggered", () => {
    expect.assertions(2);
    const wrapper = shallow(
      <Dropdown dropdownContent={() => <div className="target">test</div>}>
        {data => (
          <React.Fragment>
            <div className="click-open" onClick={data.actions.open}>
              test2
            </div>
            <div className="click-close" onClick={data.actions.close}>
              test3
            </div>
          </React.Fragment>
        )}
      </Dropdown>
    );
    wrapper.find(".click-open").simulate("click");
    expect(wrapper.find(".target")).toHaveLength(1);
    wrapper.find(".click-close").simulate("click");
    expect(wrapper.find(".target")).toHaveLength(0);
  });
  it("can remove the dropdown when triggered within the dropdown", () => {
    expect.assertions(2);
    const wrapper = shallow(
      <Dropdown
        dropdownContent={data => (
          <React.Fragment>
            <div className="target">test</div>
            <div className="click-close" onClick={data.actions.close}>
              test3
            </div>
          </React.Fragment>
        )}
      >
        {data => (
          <div className="click-open" onClick={data.actions.open}>
            test2
          </div>
        )}
      </Dropdown>
    );
    wrapper.find(".click-open").simulate("click");
    expect(wrapper.find(".target")).toHaveLength(1);
    wrapper.find(".click-close").simulate("click");
    expect(wrapper.find(".target")).toHaveLength(0);
  });

  describe("StyledDropdown", () => {
    it("renders", () => {
      mount(<StyledDropdown theme={Theme} />);
    });
  });

  describe("Item", () => {
    it("renders", () => {
      mount(<Item theme={Theme} />);
    });
  });
  describe("Spacer", () => {
    it("renders", () => {
      mount(<Spacer theme={Theme} />);
    });
  });
});
