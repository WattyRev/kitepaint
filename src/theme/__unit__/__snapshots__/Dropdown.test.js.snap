// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Dropdown renders 1`] = `
<styled.div>
  <div>
    test2
  </div>
</styled.div>
`;
